File: aeron-archive/src/main/java/io/aeron/archive/client/RecordingSignalAdapter.java
/*
 * Copyright 2014-2025 Real Logic Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.aeron.archive.client;

import io.aeron.ControlledFragmentAssembler;
import io.aeron.Subscription;
import io.aeron.archive.codecs.*;
import io.aeron.logbuffer.ControlledFragmentHandler;
import io.aeron.logbuffer.Header;
import org.agrona.DirectBuffer;

import static io.aeron.logbuffer.ControlledFragmentHandler.Action.*;

/**
 * Encapsulate the polling, decoding, and dispatching of recording transition events for a session plus the
 * asynchronous events to check for errors.
 * <p>
 * Important: set the underlying {@link RecordingSignalConsumer} instance on the {@link AeronArchive} using the
 * {@link AeronArchive.Context#recordingSignalConsumer(RecordingSignalConsumer)} method to avoid missing signals.
 *
 * @see RecordingSignal
 */
public final class RecordingSignalAdapter
{
    private final MessageHeaderDecoder messageHeaderDecoder = new MessageHeaderDecoder();
    private final ControlResponseDecoder controlResponseDecoder = new ControlResponseDecoder();
    private final RecordingSignalEventDecoder recordingSignalEventDecoder = new RecordingSignalEventDecoder();
    private final ControlledFragmentAssembler assembler = new ControlledFragmentAssembler(this::onFragment);
    private final ControlEventListener controlEventListener;
    private final RecordingSignalConsumer recordingSignalConsumer;
    private final Subscription subscription;
    private final int fragmentLimit;
    private final long controlSessionId;
    private boolean isDone = false;

    /**
     * Create an adapter for a given subscription to an archive for recording events.
     *
     * @param controlSessionId        to listen for associated asynchronous control events, such as errors.
     * @param controlEventListener    listener for control events which may indicate an error on the session.
     * @param recordingSignalConsumer consumer of recording transition events.
     * @param subscription            to poll for new events.
     * @param fragmentLimit           to apply for each polling operation.
     */
    public RecordingSignalAdapter(
        final long controlSessionId,
        final ControlEventListener controlEventListener,
        final RecordingSignalConsumer recordingSignalConsumer,
        final Subscription subscription,
        final int fragmentLimit)
    {
        this.controlSessionId = controlSessionId;
        this.controlEventListener = controlEventListener;
        this.recordingSignalConsumer = recordingSignalConsumer;
        this.subscription = subscription;
        this.fragmentLimit = fragmentLimit;
    }

    /**
     * Poll for recording transitions and dispatch them to the {@link RecordingSignalConsumer} for this instance,
     * plus check for async responses for this control session which may have an exception and dispatch to the
     * {@link ControlResponseListener}.
     *
     * @return the number of fragments read during the operation. Zero if no events are available.
     */
    public int poll()
    {
        if (isDone)
        {
            isDone = false;
        }

        return subscription.controlledPoll(assembler, fragmentLimit);
    }

    /**
     * Indicate that poll was successful and a signal or control response was received.
     *
     * @return true if a signal or control response was received.
     */
    public boolean isDone()
    {
        return isDone;
    }

    private ControlledFragmentHandler.Action onFragment(
        final DirectBuffer buffer, final int offset, final int length, final Header header)
    {
        if (isDone)
        {
            return ABORT;
        }

        messageHeaderDecoder.wrap(buffer, offset);

        final int schemaId = messageHeaderDecoder.schemaId();
        if (schemaId != MessageHeaderDecoder.SCHEMA_ID)
        {
            throw new ArchiveException("expected schemaId=" + MessageHeaderDecoder.SCHEMA_ID + ", actual=" + schemaId);
        }

        switch (messageHeaderDecoder.templateId())
        {
            case ControlResponseDecoder.TEMPLATE_ID:
                controlResponseDecoder.wrap(
                    buffer,
                    offset + MessageHeaderDecoder.ENCODED_LENGTH,
                    messageHeaderDecoder.blockLength(),
                    messageHeaderDecoder.version());

                if (controlResponseDecoder.controlSessionId() == controlSessionId)
                {
                    controlEventListener.onResponse(
                        controlSessionId,
                        controlResponseDecoder.correlationId(),
                        controlResponseDecoder.relevantId(),
                        controlResponseDecoder.code(),
                        controlResponseDecoder.errorMessage());

                    isDone = true;
                    return BREAK;
                }
                break;

            case RecordingSignalEventDecoder.TEMPLATE_ID:
                recordingSignalEventDecoder.wrap(
                    buffer,
                    offset + MessageHeaderDecoder.ENCODED_LENGTH,
                    messageHeaderDecoder.blockLength(),
                    messageHeaderDecoder.version());

                if (recordingSignalEventDecoder.controlSessionId() == controlSessionId)
                {
                    recordingSignalConsumer.onSignal(
                        recordingSignalEventDecoder.controlSessionId(),
                        recordingSignalEventDecoder.correlationId(),
                        recordingSignalEventDecoder.recordingId(),
                        recordingSignalEventDecoder.subscriptionId(),
                        recordingSignalEventDecoder.position(),
                        recordingSignalEventDecoder.signal());

                    isDone = true;
                    return BREAK;
                }
                break;
        }

        return CONTINUE;
    }
}

File: aeron-archive/src/main/cpp/client/RecordingSignalAdapter.cpp
/*
 * Copyright 2014-2025 Real Logic Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ArchiveException.h"
#include "aeron_archive_client/MessageHeader.h"
#include "aeron_archive_client/ControlResponse.h"
#include "aeron_archive_client/RecordingSignalEvent.h"
#include "RecordingSignalAdapter.h"

using namespace aeron;
using namespace aeron::archive::client;

static aeron::controlled_poll_fragment_handler_t controlHandler(RecordingSignalAdapter &adapter)
{
    return
        [&](AtomicBuffer &buffer, util::index_t offset, util::index_t length, Header &header)
        {
            return adapter.onFragment(buffer, offset, length, header);
        };
}

RecordingSignalAdapter::RecordingSignalAdapter(
    const on_control_response_t &onResponse,
    const on_recording_signal_t &onRecordingSignal,
    std::shared_ptr<aeron::Subscription> subscription,
    std::int64_t controlSessionId,
    int fragmentLimit) :
    m_fragmentAssembler(controlHandler(*this)),
    m_fragmentHandler(m_fragmentAssembler.handler()),
    m_subscription(std::move(subscription)),
    m_onResponse(onResponse),
    m_onRecordingSignal(onRecordingSignal),
    m_controlSessionId(controlSessionId),
    m_fragmentLimit(fragmentLimit)
{
}

ControlledPollAction RecordingSignalAdapter::onFragment(
    AtomicBuffer &buffer, util::index_t offset, util::index_t length, Header &header)
{
    if (m_isAbort)
    {
        return ControlledPollAction::ABORT;
    }

    MessageHeader msgHeader(
        buffer.sbeData() + offset,
        static_cast<std::uint64_t>(length),
        MessageHeader::sbeSchemaVersion());

    const std::uint16_t schemaId = msgHeader.schemaId();
    if (schemaId != MessageHeader::sbeSchemaId())
    {
        throw ArchiveException(
            "expected schemaId=" + std::to_string(MessageHeader::sbeSchemaId()) +
            ", actual=" + std::to_string(schemaId),
            SOURCEINFO);
    }

    const std::uint16_t templateId = msgHeader.templateId();
    if (ControlResponse::sbeTemplateId() == templateId)
    {
        ControlResponse response(
            buffer.sbeData() + offset + MessageHeader::encodedLength(),
            static_cast<std::uint64_t>(length) - MessageHeader::encodedLength(),
            msgHeader.blockLength(),
            msgHeader.version());

        if (response.controlSessionId() == m_controlSessionId)
        {
            m_onResponse(
                response.controlSessionId(),
                response.correlationId(),
                response.relevantId(),
                response.code(),
                response.errorMessage());
        }
    }
    else if (RecordingSignalEvent::sbeTemplateId() == templateId)
    {
        RecordingSignalEvent recordingSignalEvent(
            buffer.sbeData() + offset + MessageHeader::encodedLength(),
            static_cast<std::uint64_t>(length) - MessageHeader::encodedLength(),
            msgHeader.blockLength(),
            msgHeader.version());

        if (recordingSignalEvent.controlSessionId() == m_controlSessionId)
        {
            m_onRecordingSignal(
                recordingSignalEvent.controlSessionId(),
                recordingSignalEvent.recordingId(),
                recordingSignalEvent.subscriptionId(),
                recordingSignalEvent.position(),
                recordingSignalEvent.signalRaw());

            m_isAbort = true;
            return ControlledPollAction::BREAK;
        }
    }

    return ControlledPollAction::CONTINUE;
}
